CRAYSAT-1767: Set cray-sat version in csm-config

haasken-hpe opened the pull request:
## Summary and Scope

Add an initContainer definition to the values defined for the `csm-config` Helm chart. This initContainer sets the cray-sat image tag to match the version included in the CSM build, so that they are always consistent. The Ansible content in csm-config sets this cray-sat container image tag in the file `/opt/cray/etc/sat/version`, so that the sat-podman wrapper script knows which version of the image to run.

Add logic to `release.sh` that inspects the container image tags specified in `docker/index.yaml` and injects those tags for the cray-sat container image and the Alpine Linux container image into the csm-config Helm chart values in the Loftsman manifest.

Modify `lib/setup-nexus.sh` to set the `cray-sat` container image version and tag the latest one as `csm-latest`. We are phasing out reliance on this `csm-latest` tag, but it is still the default used by the sat-podman wrapper script if the file `/opt/cray/etc/sat/version` does not exist. Modify `lib/setup-nexus.sh` to also write the tag of the `cray-sat` image to `/opt/cray/etc/sat/version` as well. This is only executed on `ncn-m001`, but that is sufficient for bootstrapping. When the CSM layer of the CFS configuration is applied, this file will be set on all master and worker management nodes.

## Issues and Related PRs

* Resolves [CRAYSAT-1767](https://jira-pro.it.hpe.com:8443/browse/CRAYSAT-1767)
* Merge after https://github.com/Cray-HPE/csm-config/pull/213

## Testing

### Tested on:

  * mug

### Test description:

I've done various manual tests of each piece of this.

I have tested that shipping a similar Loftsman manifest on mug does result in the correct content being uploaded to the csm-config-management VCS repository. I have tested the individual `yq` commands and `sed` commands added to the `release.sh` script separately outside of their full context.

## Risks and Mitigations

This adds an initContainer to the `csm-config` Helm chart, and if that fails, the import of the csm-config-management repo to VCS will fail.

It's also possible I don't understand the build process fully, and I may have done something incorrectly that won't work due to the ordering of build steps. This should be reviewed carefully.

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

haasken-hpe committed:
CRAYSAT-1767: Set cray-sat version in csm-config

Add an initContainer definition to the values defined for the
`csm-config` Helm chart. This initContainer sets the cray-sat image tag
to match the version included in the CSM build, so that they are always
consistent. The Ansible content in csm-config sets this cray-sat
container image tag in the file `/opt/cray/etc/sat/version`, so that the
sat-podman wrapper script knows which version of the image to run.

Add logic to `release.sh` that inspects the container image tags
specified in `docker/index.yaml` and injects those tags for the cray-sat
container image and the Alpine Linux container image into the csm-config
Helm chart values in the Loftsman manifest.

Modify `lib/setup-nexus.sh` to set the `cray-sat` container image
version and tag the latest one as `csm-latest`. We are phasing out
reliance on this `csm-latest` tag, but it is still the default used
by the sat-podman wrapper script if the file `/opt/cray/etc/sat/version`
does not exist. Modify `lib/setup-nexus.sh` to also write the tag of the
`cray-sat` image to `/opt/cray/etc/sat/version` as well. This is only
executed on `ncn-m001`, but that is sufficient for bootstrapping. When
the CSM layer of the CFS configuration is applied, this file will be set
on all master and worker management nodes.

Test Description:
Only tested one piece at a time so far.

haasken-hpe commented:
I didn't realize I didn't get any automatic reviewers. Added reviewers. 

* @rbak-hpe to review the init container definition mostly
* @mtupitsyn to review the changes regarding query of container images from the docker index and other build concerns
* @heemstra to review the install changes to `lib/setup-nexus.sh` mostly
* @mharding-hpe for awareness

mtupitsyn commented on docker/index.yaml:
```diff
@@ -94,6 +94,11 @@ artifactory.algol60.net/csm-docker/stable:
+    # This image is used in an init container specified in the loftsman manifest
+    # for the csm-config import job.
+    docker.io/library/alpine:
+      - 3.18
```
Any chance just generic `alpine:3` can be used? I see it does nothing but `mkdir` and `echo`, generic image would be sufficient.

haasken-hpe replied:
Yes, it certainly could. Will that tag just resolve to whatever the latest version of the alpine image is at the time? There are probably lots of other images I could use. I just chose one that I figured would be small enough and perhaps already included so as to not inflate the size of the CSM tar file unnecessarily.

haasken-hpe replied:
@mtupitsyn, could you answer this question? Is it better for me to change this to just alpine:3 as follows?

```suggestion
      - 3
```

mtupitsyn replied:
Yep, `3` is a so called "rolling tag". Currently, rolling tags `3`, `3.18` and `latest` are all pointing to actual latest `3.18.4` at https://hub.docker.com/_/alpine.

haasken-hpe replied:
Thank you, I'll update to use just `3`.

heemstra commented on lib/setup-nexus.sh:
```diff
@@ -60,8 +60,13 @@ skopeo-sync "${ROOTDIR}/docker"
 skopeo-copy "${sat_image}:${sat_version}" "${sat_image}:csm-latest"
+# Bootstrap sat by writing the sat version to the file used by the sat-podman
+# wrapper script. This is later written by the CSM layer of the CFS config.
+echo "${sat_version}" > /opt/cray/etc/sat/version
```
The only question I have here is whether `/opt/cray/etc/sat` exists at the time this is run during a fresh install.  It'd be running on the pit node.  Does `sat` get installed there?  I didn't think so...

haasken-hpe replied:
No, to my knowledge, `sat` does not get installed on the PIT node. It would be smart to `mkdir -p /opt/cray/etc/sat` first so this command will work when the directory does not exist yet. If I fix that, I think it would be harmless in a fresh install scenario. Once the management nodes get booted with their images, they should then have `cray-sat-podman` installed, but I don't believe the CFS configuration is executed against the images before they're booted or against the nodes when they boot until later in the process. Thus, the file `/opt/cray/etc/sat/version` will not exist, and `cray-sat-podman` will default to using the `csm-latest` tag (with no caching issues because it's a fresh install), and the explicit tag will later be set by the CFS configuration. I hope that all made sense, and you can confirm.

haasken-hpe replied:
Will this be executed on the ncn-m001 node (not the PIT) during an upgrade scenario at least? That's what I gathered from [the upgrade documentation](https://github.com/Cray-HPE/docs-csm/blob/release/1.6/upgrade/Stage_0_Prerequisites.md#stage-02---prerequisites). We really only need it to bridge the gap until node personalization of the management NCNs is performed in the management-rollout stage of the IUF.

heemstra replied:
Yes - in an upgrade, it'll run on m001.  And then it'll be lost when m001 is upgraded and brought back into the cluster.  But, as you've mentioned, that's fine because CFS will replace it.

This is exactly what I had in mind: `mkdir -p /opt/cray/etc/sat`

The rest of what you said makes sense to me and jives with my understanding.

haasken-hpe replied:
Thanks for the confirmation. I think I just overlooked making sure the directory exists in this script. I will add that.

haasken-hpe committed:
CRAYSAT-1767: Ensure `/opt/cray/etc/sat` dir exists in setup-nexus.sh

Update `lib/setup-nexus.sh` to ensure the directory `/opt/cray/etc/sat`
exists before writing to the `version` file in that directory. On an
upgrade, this directory should already exist because it will have been
created by management node personalization performed by CFS. However, on
a fresh install, this script is executed on the PIT, and this directory
will not exist. In fact, in that case, `sat` is not even installed on
the PIT, so writing to this file won't affect anything, but it also
won't hurt anything.

haasken-hpe committed:
CRAYSAT-1767: Use `alpine:3` rolling tag

Use the rolling tag of `alpine:3` instead of `alpine:3.18` since we do
not depend on any particular functionality specific to a version of the
`alpine` container in its usage in the `csm-config` Helm chart's
initContainer.

haasken-hpe committed:
CRAYSAT-1767: Update to latest release of csm-config

Update the `sysmgmt.yaml` Loftsman manifest to specify the latest 1.17.0
version of the `csm-config` Helm chart. This is the newly published
release which includes the `csm.ncn.sat` role which is being modified by
the `initContainer` which is being used here.

haasken-hpe commented on docker/index.yaml:
```diff
@@ -24,7 +24,7 @@
   images:
     cray-sat:
-      - 3.25.0
+      - 3.25.5
```
Oh, I'm going to bump this to `3.25.6`, which I just released. It fixes an unrelated bug (CRAYSAT-1780).

haasken-hpe commented:
I realized last week that this is going to require one more change in order to work correctly.

Today, the SAT product stream uploads the cray-sat container image to the image registry in Nexus as `cray/cray-sat`. CSM uploads the cray-sat container image to the image registry in Nexus as `artifactory.algol60.net/sat-docker/stable/cray-sat`.

Formerly only the SAT product stream used to deliver the Ansible that caused the file `/opt/cray/etc/sat/version` to be created. Thus, if this file exists, the `sat` wrapper script provided by cray-sat-podman assumes that the image in Nexus is `cray/cray-sat`, which is where it is delivered by the separate SAT product stream. Now that the CSM product delivers the Ansible content that populates the file `/opt/cray/etc/sat/version`, we need to modify the `cray-sat-podman` wrapper script logic to always use the CSM-provided path in the image registry in Nexus.

This will create a backwards compatibility issue. If a newer version of `cray-sat-podman` is installed that has this changed logic, it will no longer be able to find the cray-sat images delivered by older versions of SAT. I think this is okay because people should not need to roll back to older cray-sat versions, and if they really need to, it would be easy enough to do by modifying the `sat` wrapper shell script or by moving images in Nexus.

Here are the steps I'm going to take to handle this:

* Change the sat builds to publish to csm-docker instead of sat-docker. The way CSM uploads container images to Nexus, it keeps the Artifactory path in the image names. I eventually planned to move SAT container images to publish to csm-docker in CRAYSAT-1771. Since I'm going to be changing the code in `sat-podman` that looks up the image in Nexus anyway, I might as well make the change to the image publishing location now, so I don't have to make two backwards incompatible changes one after the other. I will probably bump the minor version in cray-sat when I do this because if I need to patch the version of the cray-sat image included in CSM 1.5, I'm going to want to have room to bump just the patch version.
* Change the logic in the sat-podman wrapper script to only look for the cray-sat container image in the new csm-docker path in artifactory. This will probably be a major version bump, since it's incompatible with older versions of SAT.

Until I complete those changes, this should not be merged.

haasken-hpe committed:
CRAYSAT-1767: Download cray-sat image from csm-docker

The cray-sat image now publishes to csm-docker, and this is where the
newer 3.0.0 version of cray-sat-podman expects it to be uploaded to
Nexus on the system. Make the following changes:

* Download cray-sat from csm-docker. This also has the side effect of
  uploading to the same path in Nexus upon installation
* Change code in `lib/setup-nexus.sh` to tag the image as `csm-latest`
  in its new location in Nexus
* Modify the `yq` command in `release.sh` that reads the `cray-sat`
  container image version from `docker/index.yaml`
* Remove harmless but now unnecessary condition on `sat-docker` path in
  `build/images/inspect.sh`

Test Description:
Tested the `yq` command independently to ensure it could still get the
version of `cray-sat`.

haasken-hpe commented:
The actions from my previous comment here have been completed. Here are the PRs, which have been merged:

* https://github.com/Cray-HPE/sat/pull/173
* https://github.com/Cray-HPE/sat-podman/pull/35

There is still a PR which should be merged in metal-provision, but it should be timed to merge with this change, so both changes appear in the same CSM 1.6.0 build. Here is that PR:

* https://github.com/Cray-HPE/metal-provision/pull/570
